apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/account/PublicKeyServiceImpl.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright (c)  2018-2019. Apollo Foundation.
+ */
+
+package com.apollocurrency.aplwallet.apl.core.account;
+
+import com.apollocurrency.aplwallet.apl.core.app.Block;
+import com.apollocurrency.aplwallet.apl.core.app.ShufflingTransaction;
+import com.apollocurrency.aplwallet.apl.core.app.observer.events.BlockEvent;
+import com.apollocurrency.aplwallet.apl.core.app.observer.events.BlockEventType;
+import com.apollocurrency.aplwallet.apl.core.db.DbKey;
+import com.apollocurrency.aplwallet.apl.core.db.derived.EntityDbTableInterface;
+import com.apollocurrency.aplwallet.apl.core.shard.DbHotSwapConfig;
+import com.apollocurrency.aplwallet.apl.core.transaction.messages.PublicKeyAnnouncementAppendix;
+import com.apollocurrency.aplwallet.apl.core.transaction.messages.ShufflingRecipientsAttachment;
+import com.apollocurrency.aplwallet.apl.util.cache.CacheProducer;
+import com.apollocurrency.aplwallet.apl.util.cache.CacheType;
+import com.google.common.cache.Cache;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.enterprise.event.Observes;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import java.util.Arrays;
+
+@Slf4j
+@Singleton
+public class PublicKeyServiceImpl implements PublicKeyService {
+
+    @Getter
+    private final boolean cacheEnabled;
+    @Getter
+    private final Cache<DbKey, PublicKey> publicKeyCache;
+
+    private EntityDbTableInterface<PublicKey> publicKeyTable;
+    private EntityDbTableInterface<PublicKey> genesisPublicKeyTable;
+
+    @Inject
+    public PublicKeyServiceImpl(@Named("publicKeyTable") EntityDbTableInterface<PublicKey> publicKeyTable,
+                                @Named("genesisPublicKeyTable") EntityDbTableInterface<PublicKey> genesisPublicKeyTable,
+                                @Named("isPublicKeyCacheEnabled") boolean cacheEnabled,
+                                @CacheProducer @CacheType("PUBLIC_KEY_CACHE") Cache<DbKey, PublicKey> publicKeyCache ) {
+        this.publicKeyTable = publicKeyTable;
+        this.genesisPublicKeyTable = genesisPublicKeyTable;
+        this.cacheEnabled = cacheEnabled;
+        this.publicKeyCache = publicKeyCache;
+
+        log.debug("--cache-- init PUBLIC KEY CACHE={}", publicKeyCache);
+    }
+
+    void onRescanBegan(@Observes @BlockEvent(BlockEventType.RESCAN_BEGIN) Block block) {
+        clearCache();
+    }
+
+    void onDbHotSwapBegin(@Observes DbHotSwapConfig dbHotSwapConfig) {
+        clearCache();
+    }
+
+    void onBlockPopped(@Observes @BlockEvent(BlockEventType.BLOCK_POPPED) Block block) {
```

Comment:
We've got advantage by using publicKey entity for caching. So that we can react on blockchain.rollback by removing all entries with the height greater than rollback height (you've alredy implemented this). Therefore such event observer may be completely removed (additional testing required)

Reply:
Dear colleagues! Please consider that we planned already full re-design of DB access level as a part of work on other big task in Q2. So in my opinion let's accept this solution because it solves minor problem now.